src/modules/main/PlayerView.tsx

```diff
@@ -1,23 +1,159 @@
-import { Box, Typography } from '@mui/material';
+import { useEffect, useState } from 'react';
 
-import { useLocalContext } from '@graasp/apps-query-client';
+import {
+  Alert,
+  Box,
+  Button,
+  CircularProgress,
+  Container,
+  Stack,
+  Typography,
+} from '@mui/material';
 
-import { hooks } from '@/config/queryClient';
+import isEqual from 'lodash.isequal';
+
+import { AnsweredLabel, Label, Settings, SettingsKeys } from '@/@types';
+import { ANSWER_SUBMISSION_TYPE } from '@/config/constants';
+import { useAppTranslation } from '@/config/i18n';
+import { hooks, mutations } from '@/config/queryClient';
 import { PLAYER_VIEW_CY } from '@/config/selectors';
+import { APP } from '@/langs/constants';
+
+import PlayerFrame from '../common/PlayerFrame';
+
+type SubmittedAnswer = {
+  expectedId: string;
+  actualId?: string;
+};
 
 const PlayerView = (): JSX.Element => {
-  const { permission } = useLocalContext();
   const { data: appContext } = hooks.useAppContext();
-  const members = appContext?.members;
+  const { t } = useAppTranslation();
+  const { data: appData } = hooks.useAppData<{ answers: SubmittedAnswer[] }>();
+  const { data: appSettings, isLoading } = hooks.useAppSettings<Settings>({
+    name: SettingsKeys.SettingsData,
+  });
+  const { mutate: saveAppData } = mutations.usePostAppData();
+  const [answeredLabels, setAnsweredLabels] = useState<AnsweredLabel[]>([]);
+  // labels will be null only before setting the state as we cannot render all labels within container if not settled yet
+  const [labels, setLabels] = useState<null | Label[]>(null);
+
+  const { data: image } = hooks.useAppSettings({
+    name: SettingsKeys.File,
+  });
+  const [isSubmitted, setIsSubmitted] = useState(false);
+
+  const answersAppData = appData?.filter(
+    ({ type }) => type === ANSWER_SUBMISSION_TYPE,
+  );
+  const settingLabels = appSettings?.[0].data.labels;
+  const lastVersion = answersAppData?.[answersAppData.length - 1];
+  const answers = lastVersion?.data.answers;
+
+  const retry = (): void => {
+    if (settingLabels) {
+      const answered = settingLabels?.map((label) => ({
+        expected: label,
+        actual: null,
+      }));
+
+      setAnsweredLabels(answered);
+      setLabels(settingLabels);
+    }
+  };
+
+  const save = (): void => {
+    const submittedAnswers = answeredLabels.map(({ expected, actual }) => ({
+      expectedId: expected.id,
+      actualId: actual?.id,
+    }));
+    saveAppData({
+      data: { answers: submittedAnswers },
+      type: ANSWER_SUBMISSION_TYPE,
+    });
+  };
+
+  useEffect(() => {
+    if (settingLabels) {
+      if (answers) {
+        const answered = answers?.map(
+          ({ expectedId, actualId }: SubmittedAnswer) => ({
+            expected: settingLabels.find(
+              ({ id }) => id === expectedId,
+            ) as Label,
```

Comment:
Why do you have to cast the expected but not the actual?

Reply:
because `expected`  always should be defined, As I'm doing `appSettings.find` TS expect that it might be undefined

Reply:
Okay thanks for the explanations. What do you think about using a utils function like:

```ts
// I don't know if the name is clear enough.
const ensure = <T>(value: T, name: string = 'value') => {
  if (!value) {
    throw new TypeError(`The ${name} must be defined!`);
  }

  return value;
}
```

In this way, you ensure that the found value is defined and you don't have to cast.

```ts
({ expectedId, actualId }: SubmittedAnswer) => ({
   expected: ensure(settingLabels.find(
      ({ id }) => id === expectedId,
   ), 'expected'),
})
```